### ISSUE TITLE
Merge simuvex and angr

### ISSUE THREAD

**Issue by zardus** [angr/angr#349]
# Current Status

SimuVEX is no more. It's been absorbed into angr, and lives on as a compatibility stub. Path is gone, and SimulationManager is the new PathGroup. The future is here.

# The Plan

The current plan is to take this in 3 phases:

* [x] Phase 1: merge simuvex into angr, redistribute the files under the angr module. No functionality changes.
  * [x] angr passes
  * [x] angr-bf passes
  * [x] angr-doc passes
  * [x] tracer passes
  * [x] rex passes
  * [x] colorguard passes
  * [x] identifier passes
  * [x] fidget passes
  * [x] patcherex passes
  * [x] angrop passes
* [x] Phase 2: remove Path, make PathHistory a SimStatePlugin called StateHistory, rename PathHierarchy to StateHierarchy, rename PathGroup to SimContext.
  * [x] angr passes
  * [x] angr-bf passes
  * [x] angr-doc passes
  * [x] tracer passes
  * [ ] rex passes
  * [x] colorguard passes
  * [x] identifier passes
  * [x] fidget passes
  * [x] patcherex passes
  * [x] angrop passes
* [x] Phase 3: full backwards compatibility -- stub modules for `simuvex`, `angr.Path`, and alias for `angr.PathGroup`
  * [x] simuvex stub
  * [ ] pass-throughs for state.log and state.scratch
  * [x] alias for PathGroup
  * [x] alias for `state.state`
  * [ ] alias for `state.length`
  * [ ] old angr-doc examples pass
  * [ ] old angr testcases pass
  * [ ] old angr-bf testcases pass
  * [ ] old tracer passes
  * [ ] old rex passes
  * [ ] old colorguard passes
  * [ ] old identifier passes
  * [ ] old fidget passes
  * [ ] old patcherex passes
  * [ ] old angrop passes
* [ ] Phase 4: clean up remaining issues in the SimuVEX repo.
* [ ] Phase 5: squash angr commits to make Fish happy

# Original Musings

@rhelmot and I had a discussion that started out with my frustration that adding a syscall to angr requires two PRs (one to SimuVEX for the SimProcedure and one for angr to change the SimOS syscall table). @rhelmot made an unorthodox suggestion that I'll reiterate toward the end of this issue. First, some observations:

1. Some of the modules we've written for angr are being used in other projects. This includes CLE, PyVEX, claripy, and archinfo. One module conspicuously missing from this list (as far as I know) is SimuVEX.
2. This is probably because the other modules have very well-defined, self-contained purposes, separate from angr. SimuVEX, on the other hand, lives as a tightly-coupled state mutation engine for angr (you can probably begin to see where this is going now).
3. On the angr side, over the last year, `Path` has become and more and more thin wrapper around `SimState`, `PathHistory`, and `CallStack` (the last of which really belongs in `PathHistory`, anyways), and convenience accessors for those (i.e., `path.events`, `path.actions`, `path.jumpkind`, and the like are all pass-throughs to data stored in `PathHistory`).
4. State options, while technically a SimuVEX construct and technically tied to the state, is used to control behavior all over the place in angr (i.e., in the CFG, in `PathHistory`, in `PathHierarchy`, etc).
5. SimInspect breakpoints, also technically a SimuVEX construct, have all sorts of angr-triggered events.
6. We're constantly running into problems where we don't have access to the path or project from SimuVEX (specifically, from `SimInspect` hooks).
7. Because we run into the same problem tracking information over successive paths as we do over successive states, we've invented `path.info`, which is basically a stripped down `SimStatePlugin`.
8. We've had debates on, for example, pulling SimProcedures out of SimuVEX, pulling SimOS out of angr and into some weird joint package with the SimProcedures, etc.
9. SimEngines are currently split across two different packages: `SimEngineVEX`, `SimEngineProcedure`, and `SimEngineUnicorn` are in SimuVEX, `SimEngineFailure` and `SimEngineSyscall` are in angr.
10. Simu**VEX** is a weird name for a state mutation engine that supports multiple execution/translation engines.

Basically, the decoupling between angr and simuvex is not really there, unlike it is with CLE, PyVEX, claripy, archinfo, etc. A proposed solution is to **eliminate simuvex itself, and merge it into angr**. The changes would be:

1. SimuVEX would be merged into the angr repo.
2. `Path` would be eliminated. Each `SimState` would have a `state.history` attribute with the `PathHistory` (now probably called `StateHistory`). This would eliminate the weird thing where we copy in everything from the `state.log` plugin into `path.history` every state -- they could be the same entity.
3. `PathGroup` would now wrangle stashes of states instead of stashes of paths.

To make the transition smoother, we'd have the following deprecation support:

1. If we create a `state.state` property, people can still do stuff like `path_groups.found[0].state` as before. Likewise for properties for `state.history` (which would be the replacement of `state.log`, anyways), `state.callstack`, and so on.
2. We'd probably have to leave a dummy simuvex package for a while, with deprecated passthroughs to the old concepts.

This is a big conceptual change -- it's angr 7.7.4.1 sort of stuff for sure, but it seems to me that it would make sense and would simplify using and extending angr.

**Comment by ltfish** [comment]
I'm down for this change. I see it as an improvement. The separation between angr and SimuVEX is not necessary, as people are not able to use SimuVEX without angr anyways. Comparing the number of commits and the frequency of commits we make to angr/SimuVEX combined to those of claripy can show why this can be a good change.

However, such a transition is huge, and the resulting code base can take a long time to stabilize. My suggestion is we sit on this idea for at least a month, and think over it as much as possible, then take a week off of our normal schedule to finish everything together.

**Comment by ltfish** [comment]
As a side note: I'm against blowing away `Path`. `Path` is still useful as a descriptor of what addresses are come across, what events have happened, and what states there are during an execution. It seems to me that using `state.history` for that purpose is an abuse.

**Comment by Lukas-Dresel** [comment]
Hey, so I have only recently come in contact with this issue, but I support fish in saying that I am against blowing away Path. Instead I would propose a clearer distinction between states and Paths. Here is an excerpt of things that I jotted down during a bout of insomnia last night. 

Take this with a grain of salt, it's only points I jotted down so I would remember it once I woke back up ;)

```
TL;DR: Move states into angr, make paths logically be the path how to get to a particular state, make Simuvex an engine. 
Process-state related to but not equal to environment-state.


1. Make sure all the information that is actually path information is tracked in the path!

 - no more constraints, log, previous_run, etc. In the states!
 - easier deduplication of multiple states
 - possible to split paths on memory write ( this is currently handled solely in simuvex, 
   we could make it possible for a path to specify which it wants and why)      
 - paths without states, states without paths possible ( possible to prune states of paths 
   we don't care about any more while keeping trace, guards, etc.) (States without paths as 
   it's currently possible of course)


2. Rather than merging angr and simuvex, make angr use and expose all the simuvex magic 
and coolness in the paths! ( PATH-INSPECT-BREAKPOINTS, path solving, actions, log, events, guards, etc. )


3. States can never be unsatisfiable! They can only be unreachable. Keeping states seperate 
and having the constraints in the paths makes this clear. I realized this from fish's constant 
corrections about the difference between ASTs and Constraints which I incorrectly kept using 
interchangeably. States have ASTs, paths have constraints! This distinction is important.


4. Move engines to paths => hit JNI? SOOT engine => VEX engine.
Similar to paths, set initial engine during initialization, everything else happens based on 
triggers, e.g. path breakpoint can switch engine, switch state, set satisfiability, 
introduce constraints, etc.


5. Move hooks and inspect-breakpoints to paths. That way angr-gdb is gonna be way cooler! Set the 
hooks on the initial path, when something special happens => remove them and install different hook

That way we also get rid of the dependence on simuvex for breakpoints. Breakpoints are neat and cool 
=> they should be available for other stuff

Imagine someone writes a bf engine that also exposes instruction, mem_read and mem_write hooks => boom 
you get tracking that for free if you hook 'mem_read'. Oh, you wanted only simuvex memory reads? Hook 
'simuvex.mem_read instead' (engine names?)


6. Simprocedures to paths => simply different engine, it's free! Engines would have to be changed 
slightly to return successor states as well as additional path changes, e.g added constraints for 
each successor, etc.


5 + 6 solves weirdness of passing simuvex.SimProcedure to angr.Project for a hook? The project as I 
understand it doesn't necessarily have to do with execution. Have it on the path instead! Paths are 
made to track and perform execution


6. Have engine information stored on the path: unicorn concretization limits, vex simplification level, 
how to lift for this engine (bytes from binary vs. Bytes from state, vex from bytes, vex from program, 
only operates on bytes directly (brainfuck engine for example)), etc.

vex special function python implementations (ccalls or whatever they are called) maybe too?


Possibly:

    loaded library engine => base_addr + the mapping of addr to summary simprocedure
    syscall engine => syscall num to simprocedure mapping, 

Both simply calling the real simprocedure engine internally, or they can also be part of the 
simprocedure engine


The really cool thing: This in combination with a cool lifter pipeline ( see my BB changes ) allows us to do 
stuff like 

 - load bytes from program until a path has detected a jump to written memory => switch to loading bytes from state
 - Only decode SOOT => JNI call => decode vex for simuvex => return from JNI => back to decoding SOOT


7. States are the state of the process not the state of the environment. 

Process state: registers, memory, held locks, open file descriptors, open network connections, current MM I/O mapping, etc. => on State!
Environment state: file contents, network communications, I/O interactions, loaded libraries, etc. => on path

8. Paths with multiple states: one java state, one native state
```

**Comment by zardus** [comment]
After some further discussion we're currently considering the following proposal:

## The Current Proposal

- `Path` will be eliminated
- `SimStateLog`, `SimStateScratch`, and `PathHistory` will be integrated into the `state.history` plugin.
- `PathGroup` will be renamed to _something_ (current ideas are `ExplorationContext` or `Executor`).
- the executor will contain stashes of `SimState` objects.
- for convenience, as a replacement for `Path.step` (which is apparently used by people), we will have `state.step()`, which will instantiate an execution, do a single step, and return the new executor.


There's some idea to implement this over the course of the next week (though first we have to merge some outstanding branches). However, this is still all very much open to further ideas.

**Comment by ltfish** [comment]
To reiterate, major changes will happen over the next week.

**Comment by zardus** [comment]
Okay, the first phase of this work is nearly complete: simuvex has been merged into the angr repo in the `wip/the_end_times` branch. Be afraid.

**Comment by ltfish** [comment]
Squash the commits!

**Comment by zardus** [comment]
Squashing the commits will be Phase 4 :-)

**Comment by ltfish** [comment]
I propose we properly squash commits at the end of each phase, so we can revert individual phases.

**Comment by axt** [comment]
Good luck for the merge :).

Just a note: I didn't checked this more deep, but github says this about the branch, and I think that can't be right: 'This branch is 3430 commits ahead of master.'

**Comment by zardus** [comment]
We're just crazy like that:

```
git pull git@github.com:angr/simuvex
```

**Comment by axt** [comment]
@zardus Ah, sorry for the noise then. I thought its some git branch-splitting issue.

**Comment by zardus** [comment]
I'm starting in on phase 4 -- closing out old simuvex PRs and issues.

**Comment by rhelmot** [comment]
I banish this issue to the depths of hell
